common/namespace/registry.go

```diff
@@ -456,25 +496,44 @@ UpdateLoop:
 		r.cacheNameToID = newCacheNameToID
 		return oldEntries, newEntries
 	})
+
+	r.callStateChangeCallbacks(stateChanged, allEntries)
+
 	return nil
 }
 
-func (r *registry) updateNameToIDCache(c cache.Cache, name Name, id ID) {
-	c.Put(name, id)
+func (r *registry) callStateChangeCallbacks(stateChanged, allEntries []*Namespace) {
+	r.stateChangeCallbacksLock.Lock()
+	defer r.stateChangeCallbacksLock.Unlock()
```

Comment:
is there a way to avoid holding this lock while making all the callbacks? this could lead to deadlock issue if not handled very very carefully.

Reply:
On second thought, I think I should leave it how it was at first: the lock in question here is only protecting these two maps. The only other place it's used is registering/unregistering callbacks, which should only be called at startup/shutdown. So I think it's fine to call while holding this lock and doesn't warrant any more complexity.